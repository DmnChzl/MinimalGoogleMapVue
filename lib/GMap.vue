<template>
  <div id="mnml-gmap" ref="gMapRef" />
</template>

<script lang="ts">
import { defineComponent, PropType, ref, onMounted } from 'vue';

interface LatLng {
  lat: number;
  lng: number;
  onClick?: () => void;
}

interface ColorStyler {
  color: string;
}

interface LightStyler {
  lightness: number;
}

interface SaturationStyler {
  saturation: number;
}

interface VisibilityStyler {
  visibility: string;
}

interface WeightStyler {
  weight: number;
}

type Styler = ColorStyler | LightStyler | SaturationStyler | VisibilityStyler | WeightStyler;

interface StyledMap {
  featureType?: string;
  elementType?: string;
  stylers: Styler[];
}

export default defineComponent({
  name: 'GMap',
  props: {
    apiKey: String,
    defaultCenter: {
      type: Object as PropType<LatLng>,
      default: () => ({ lat: 0, lng: 0 })
    },
    defaultZoom: {
      type: Number,
      default: 10
    },
    enableUI: {
      type: Boolean,
      default: false
    },
    markers: {
      type: Array as PropType<LatLng[]>,
      default: () => []
    },
    styledMap: {
      type: Array as PropType<StyledMap[]>,
      default: () => []
    }
  },
  setup(props) {
    const gMapRef = ref(null);

    const createMap = (divElement: HTMLDivElement) => {
      const map = new google.maps.Map(divElement, {
        center: {
          lat: props.defaultCenter.lat,
          lng: props.defaultCenter.lng
        },
        zoom: props.defaultZoom,
        disableDefaultUI: !props.enableUI
      });

      if (props.styledMap.length > 0) {
        const styledMapType = new google.maps.StyledMapType(props.styledMap);
        map.mapTypes.set('styled_map', styledMapType);
        map.setMapTypeId('styled_map');
      }

      return map;
    };

    // prettier-ignore
    const initMarker = (map: google.maps.Map) => ({ lat, lng }: LatLng) => {
      return new google.maps.Marker({
        position: { lat, lng },
        map
      });
    };

    onMounted(() => {
      if (props.apiKey) {
        const mapScript = document.createElement('script');
        mapScript.src = `https://maps.googleapis.com/maps/api/js?key=${props.apiKey}`;
        mapScript.async = true;
        window.document.body.appendChild(mapScript);

        mapScript.addEventListener('load', () => {
          if (gMapRef.value !== null) {
            const gMap = createMap(gMapRef.value);
            const createMarker = initMarker(gMap);

            props.markers.forEach(({ lat, lng, onClick }) => {
              const marker = createMarker({ lat, lng });

              if (onClick) {
                marker.addListener('click', onClick);
              }
            });
          }
        });
      }

      if (gMapRef.value !== null) {
        gMapRef.value.style.height = '100%';
      }
    });

    return {
      gMapRef
    };
  }
});
</script>
