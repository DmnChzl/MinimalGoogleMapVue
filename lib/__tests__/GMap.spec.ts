import { render } from '@testing-library/vue';
import GMap from '../GMap.vue';

const API_KEY = 'YLzo6G7hJxAg55x/EvkEaPTvoO/PhQ9ASxzZ5A6OsNJLE4dxq+zK2lJ5Ta7/z5bbtQ9C2f2jDRRW6JyUArVIpQ==';

describe('<GMap />', () => {
  it('Should Match Snapshot', () => {
    const { container } = render(GMap, { props: { apiKey: API_KEY } });
    expect(container).toMatchSnapshot();
  });

  it('Should Renders MNML Element', () => {
    const { container } = render(GMap, { props: { apiKey: API_KEY } });
    expect(container.querySelector('#mnml-gmap')).toBeInTheDocument();
  });
});
