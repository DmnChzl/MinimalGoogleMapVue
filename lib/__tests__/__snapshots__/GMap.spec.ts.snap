// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GMap /> Should Match Snapshot 1`] = `
<div>
  <div
    id="mnml-gmap"
  />
</div>
`;
